In [2]:
import sys
import os

sys.path.append(os.path.abspath("../Src"))

from log.utils.logger import setup_logger

logger.py location: e:\Practicum Project\Practicum\Src\log\utils\logger.py
Log directory: e:\Practicum Project\Practicum\Src\log\Log_Data


In [3]:
logger = setup_logger("11_GridSearch")
logger.info("Starting normalization pipeline...")

2026-07-02 20:32:59 | INFO | Starting normalization pipeline...


In [4]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd

from sklearn.model_selection import GridSearchCV
from sklearn.metrics import r2_score, mean_squared_error

# Regression Models
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import (
    RandomForestRegressor,
    GradientBoostingRegressor,
    ExtraTreesRegressor,
    AdaBoostRegressor
)
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from sklearn.neural_network import MLPRegressor

try:
    from xgboost import XGBRegressor
    xgb_available = True
except:
    xgb_available = False

In [5]:
X_train = pd.read_csv("../Data/Processed/X_train.csv")
X_test = pd.read_csv("../Data/Processed/X_test.csv")

y_train = pd.read_csv("../Data/Processed/y_train.csv").squeeze()
y_test = pd.read_csv("../Data/Processed/y_test.csv").squeeze()

Model Dictonary

In [6]:
models = {

    "Linear Regression": (
        LinearRegression(),
        {}
    ),

    "Ridge": (
        Ridge(),
        {
            "alpha":[0.1,1,10]
        }
    ),

    "Lasso": (
        Lasso(),
        {
            "alpha":[0.001,0.01,0.1]
        }
    ),

    "Decision Tree": (
        DecisionTreeRegressor(random_state=42),
        {
            "max_depth":[5,10,None],
            "min_samples_split":[2,5]
        }
    ),

    "Random Forest": (
        RandomForestRegressor(random_state=42),
        {
            "n_estimators":[100,200],
            "max_depth":[5,10,None]
        }
    ),

    "Gradient Boosting": (
        GradientBoostingRegressor(random_state=42),
        {
            "n_estimators":[100,200],
            "learning_rate":[0.01,0.1]
        }
    ),

    "Extra Trees": (
        ExtraTreesRegressor(random_state=42),
        {
            "n_estimators":[100,200]
        }
    ),

    "AdaBoost": (
        AdaBoostRegressor(random_state=42),
        {
            "n_estimators":[50,100]
        }
    ),

    "KNN": (
        KNeighborsRegressor(),
        {
            "n_neighbors":[3,5,7]
        }
    ),

    "SVR": (
        SVR(),
        {
            "C":[1,10],
            "kernel":["linear","rbf"]
        }
    ),

    "MLP": (
        MLPRegressor(random_state=42,max_iter=1000),
        {
            "hidden_layer_sizes":[(50,),(100,)],
            "alpha":[0.0001,0.001]
        }
    )
}

In [7]:
#optional  xgb boost
if xgb_available:

    models["XGBoost"] = (

        XGBRegressor(
            random_state=42,
            objective="reg:squarederror"
        ),

        {
            "n_estimators":[100,200],
            "max_depth":[3,5],
            "learning_rate":[0.05,0.1]
        }
    )

In [8]:

logger.info("Initializing Grid Search Loop for model Selection...")
results = []

best_model = None
best_score = -999999
best_name = ""

2026-07-02 20:33:29 | INFO | Initializing Grid Search Loop for model Selection...


In [9]:
logger.info("Started Grid Search Loop for model Selection...")
for name, (model, params) in models.items():

    print("="*60)
    print(name)

    grid = GridSearchCV(

        estimator=model,

        param_grid=params,

        cv=5,

        scoring="r2",

        n_jobs=-1

    )

    grid.fit(X_train, y_train)

    y_pred = grid.predict(X_test)

    r2 = r2_score(y_test, y_pred)

    rmse = mean_squared_error(
        y_test,
        y_pred
    ) ** 0.5

    results.append({

        "Model":name,

        "Best CV Score":grid.best_score_,

        "Test R2":r2,

        "RMSE":rmse,

        "Best Params":grid.best_params_

    })

    if r2 > best_score:

        best_score = r2
        best_model = grid.best_estimator_
        best_name = name

    print("Best CV :",grid.best_score_)
    print("Test R2 :",r2)
    print("RMSE :",rmse)
    logger.info("Completed Grid Search Loop for model Selection...")

2026-07-02 20:33:29 | INFO | Started Grid Search Loop for model Selection...


Linear Regression


2026-07-02 20:33:44 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.4435098330624195
Test R2 : 0.46900720167493604
RMSE : 0.727351882277484
Ridge


2026-07-02 20:33:49 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.4435243982880829
Test R2 : 0.4689899584976186
RMSE : 0.727363692000422
Lasso


2026-07-02 20:33:50 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.4434432248415902
Test R2 : 0.46932062382686623
RMSE : 0.7271371883656708
Decision Tree


2026-07-02 20:33:51 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.4274271713386219
Test R2 : 0.45517470880090194
RMSE : 0.7367648216802991
Random Forest


2026-07-02 20:35:05 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.47880773872387633
Test R2 : 0.5051336116468075
RMSE : 0.7021731686140885
Gradient Boosting


2026-07-02 20:35:33 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.48254957414118865
Test R2 : 0.5156637258446535
RMSE : 0.6946623319053176
Extra Trees


2026-07-02 20:36:01 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.43132166774377456
Test R2 : 0.4716711804240685
RMSE : 0.7255250341869076
AdaBoost


2026-07-02 20:36:06 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.4075786067832695
Test R2 : 0.43484424186266546
RMSE : 0.7503853213294653
KNN


2026-07-02 20:36:08 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.42915169215837984
Test R2 : 0.44489726602479185
RMSE : 0.7436814259270498
SVR


2026-07-02 20:39:45 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.4420251937907193
Test R2 : 0.4712950990664502
RMSE : 0.7257832142161086
MLP


2026-07-02 20:40:30 | INFO | Completed Grid Search Loop for model Selection...


Best CV : 0.4697762541018591
Test R2 : 0.5000988901808748
RMSE : 0.7057360494731479


In [10]:
logger.info("Results of Grid Search Loop for model Selection...")
results_df = pd.DataFrame(results)

results_df = results_df.sort_values(
    by="Test R2",
    ascending=False
)

results_df

2026-07-02 20:40:30 | INFO | Results of Grid Search Loop for model Selection...


,Model,Best CV Score,Test R2,RMSE,Best Params
5,Gradient Boosting,0.482550,0.515664,0.694662,"{'learning_rate': 0.1, 'n_estimators': 200}"
4,Random Forest,0.478808,0.505134,0.702173,"{'max_depth': 10, 'n_estimators': 200}"
10,MLP,0.469776,0.500099,0.705736,"{'alpha': 0.001, 'hidden_layer_sizes': (50,)}"
6,Extra Trees,0.431322,0.471671,0.725525,{'n_estimators': 200}
9,SVR,0.442025,0.471295,0.725783,"{'C': 1, 'kernel': 'rbf'}"
2,Lasso,0.443443,0.469321,0.727137,{'alpha': 0.001}
0,Linear Regression,0.443510,0.469007,0.727352,{}
1,Ridge,0.443524,0.468990,0.727364,{'alpha': 10}
3,Decision Tree,0.427427,0.455175,0.736765,"{'max_depth': 5, 'min_samples_split': 5}"
8,KNN,0.429152,0.444897,0.743681,{'n_neighbors': 7}


In [11]:
print("Best Model :",best_name)
print("Best Test R2 :",best_score)
logger.info(f"Best Model : {best_name}")
logger.info(f"Best Test R2 : {best_score}")

2026-07-02 20:40:30 | INFO | Best Model : Gradient Boosting
2026-07-02 20:40:30 | INFO | Best Test R2 : 0.5156637258446535


Best Model : Gradient Boosting
Best Test R2 : 0.5156637258446535


In [12]:
import joblib

joblib.dump(

    best_model,

    "../model/best_regression_model.pkl"

)

['../model/best_regression_model.pkl']

Gradient Boosting Regressor

Cross-validation R²: 0.483
Test R²: 0.516
RMSE: 0.695

This means the model explains approximately 52% of the variation in fatigue scores using the available features.

For a real-world sports science dataset—with missing values, subjective wellness measures, and biological variability—an R² around 0.5 is generally respectable. Fatigue is influenced by many factors that are not captured in the dataset (nutrition, recovery quality, stress outside training, illness, etc.), so it's uncommon to achieve extremely high R² values.

Why Gradient Boosting won

Gradient Boosting is well suited because:

It captures nonlinear relationships.
It models interactions between variables without manual feature engineering.
It is robust to mixed feature types after preprocessing.
It often performs well on structured/tabular datasets like yours.